cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -405,7 +434,33 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
 		return err
 	}
 
-	// 4.1. read modeenv
+	// 4.1 verify that ubuntu-data comes from where we expect it to
+
+	// since ubuntu-data has already been mounted, see if there is an encrypted
+	// device that we should try to boot from
+	diskOpts := &disks.Options{}
+	_, err = disk.FindMatchingPartitionUUID("ubuntu-data-enc")
+	if err == nil {
+		// we found an ubuntu-data-enc device on the disk we mounted ubuntu-boot
+		// from, so assume that ubuntu-data is a decrypted mapper device back to
+		// ubuntu-data-enc
+		diskOpts.IsDecryptedDevice = true
+	}
+
+	matches, err := disk.MountPointIsFromDisk(boot.InitramfsDataDir, diskOpts)
+	if err != nil {
+		return err
+	}
+	if !matches {
+		// failed to verify that ubuntu-data mountpoint comes from the same disk
+		// as ubuntu-boot
+		if diskOpts.IsDecryptedDevice {
+			return fmt.Errorf("cannot validate boot: ubuntu-data mountpoint is not from ubuntu-data-enc partition")
```

Comment:
I wonder if we should provide more error information in this case. Something like: `cannot validate boot: ubuntu-data mountpoint is from disk /dev/sda1 but from ubuntu-data-enc partition (/dev/sdb2)` or something like this? Only if we have the data of course. But giving some more hints is probably a good idea(?).

Reply:
I will reword this to be:

```
cannot validate boot: ubuntu-data mountpoint is expected to be from disk 252:1 but is not
```

because we don't currently have a way to get the /dev node for a disks.Disk, but we do have Dev() which returns the major/minor number. I like your suggestion though, and I think it would be easy to add this to the disks package but I'd rather do that in a followup to show /dev node instead of the device major/minor number. We also don't currently have a way to tell what disk a mount point is from at this point in the case they don't match, I think that probably we could modify the function signature to return what disk the mount point is from if it was found however that would also be a followup IMHO.

Reply:
I added some TODOs and reworded the message, let me know if you want it reworded further